### ISSUE TITLE
Validation retriggers when component is mounted again.

### ISSUE THREAD

**Issue by Nicolas-Medina-Teracode** [vercel/swr#597]
# Bug report

Suppose I need to fetch data with a POST request, where I have to send the ids in the request body.

```
const MyComponent = (props) => {
    const body = useMemo( () => ({ ids: props.ids }), []) 
    const { data } = useSWR(['/api', body])

    return //...
}
```
This works fine, the problem is that when I go to another page and then get back to the first one (containing MyComponent) the request is retriggered. This makes sense since useMemo only caches the data as long as the component is mounted. But is there a way to avoid retriggering the request when mounting the component again?

Thanks in advance,

**Comment by sergiodxa** [comment]
You can set `revalidateOnMount` as false in the hook options. But ideally you use SWR only for GET requests, because the idea of SWR is that it will revalidate in many cases to ensure data consistency

**Comment by Nicolas-Medina-Teracode** [comment]
Thanks for your answer @sergiodxa!! 

I was wondering if there is no way to define the key as a fixed string (so SWR would know how to handle the cache and revalidation) and pass the body as a separate param that would not be compared.

I thinking about this because I have another use case where I am trying to update a chart with the data of the last 10 minutes.
In this case, my endpoint URL is always the same `/chart-data` but what is changing are the query params that I am sending `/chart-data?from=123&to=456`, where from and to should be the time range from the last 10 minutes.

I was reading that there is a way to implement a custom cache, but I don't know if that would be an overkill and I should not use SWR at all in this case.

Thanks again man!

**Comment by sergiodxa** [comment]
The custom cache is not yet a thing, what I would do in your case is to prefetch the data before changing the key in SWR, this way once you change it is already there.

**Comment by Ashot-KR** [comment]
> You can set `revalidateOnMount` as false in the hook options. But ideally you use SWR only for GET requests, because the idea of SWR is that it will revalidate in many cases to ensure data consistency

with `revalidateOnMount: false` swr doesn't make request even when there is no cached data yet.
with `revalidateOnMount: true` request will be send even if cached date exists.
I this option `revalidateOnCacheExists: true`(or something like this) will be useful when we need to display cached date without any data requests.

**Comment by nikhilag** [comment]
What's the reason for not fetching data if it doesn't exist in cache when we set `revalidateOnMount: false`?

**Comment by haluvibe** [comment]
```// Fetch API (optional)
  useEffect(() => {
    if (debouncedId) {
      getResource(debouncedId)
    }
    if (debouncedId2) {
      getResource(debouncedId2)
    }
  }, [debouncedId, debouncedId2])

  return (
    <>
      <label htmlFor="id">Resource ID:</label>{' '}
      <input id="id" type="text" onChange={handleChange} value={id} />
      <input id="id2" type="text" onChange={handleChange2} value={id2} />
      {!debouncedId || !debouncedId2 ? (
        <p>Enter IDs</p>
      ) : (
        <>
          <Resource id={debouncedId} />
          <Resource id={debouncedId2} />
        </>
      )}
    </>
  )
}```

**Comment by haluvibe** [comment]
```export const Resource: React.FunctionComponent<IResourceProps> = ({ id }) => {
  const { data } = useResource(id)
  if (!data) return <Loading />
  return <p>{data.url}</p>
}```

**Comment by haluvibe** [comment]
```
import useSWR, { ConfigInterface, cache } from 'swr'
import { fetchAndCache, fetcher } from '../utils/swr.utils'

export const getResource = async (id: string) => {
  if (!id) return
  return fetchAndCache('resource/' + id)
}

export const useResource = (id: string, config: ConfigInterface = {}) => {
  const currentData = cache.get(['resource', id])
  console.log('currentData', currentData)
  return useSWR('resource/' + id, fetcher, { revalidateOnMount: false, ...config })
}
```

**Comment by gcor** [comment]
There is the [deduplication option](https://swr.vercel.app/advanced/performance#deduplication) that helps to prevents unnecessary requests.

**Comment by danharper** [comment]
> What's the reason for not fetching data if it doesn't exist in cache when we set revalidateOnMount: false?

Yeah I found this weird too!

e.g. in my app I'm pre-populating individual item keys (e.g. `/api/locations/1`, `/api/locations/2`) when a "list" request is made, e.g. `/api/locations`), and don't need to send that `/api/locations/1` request when you go to the details page; using SWR's `mutate` function.

I can set `revalidateOnMount: false` to avoid the useless network request; but I want it to be made if I forgot to populate the cache first!

So, I'm using a workaround to first check in the cache, then send a network request. (To workaround the "rule of hooks", it's using useSWR(null) if the cache lookup returned; and it relies on the cache lookup being sync -- idk if that could ever change)

```ts
// useCachingSWR.ts
import useSWR, {ConfigInterface, responseInterface} from 'swr';

export default function useCachingSWR<Data, Error = any>(
  key: string,
  options?: ConfigInterface<Data, Error>,
): responseInterface<Data, Error> {
  const fromCache = useSWR<Data, Error>(key, {...options, revalidateOnMount: false});
  const fromNetwork = useSWR(fromCache.data ? null : key, options);
  return fromCache.data ? fromCache : fromNetwork;
}
```

now, if you: `const locationLoader = useCachingSWR<Location>('/api/locations/'+locationId)`, the request won't be made if you've already cached it.


(No idea if there's a better way of doing this; I'm new to SWR)

**Comment by shuding** [comment]
I strongly disagree that those are "useless network request", I want to clarify the idea behind this.

The name of SWR stands for "stale-while-revalidate". Which means "even if we have the stale cache, we still send the request to validate that our cache is up to date". SWR isn't simply just a cache, what it provides is better UX and **eventually data consistency**. The revalidation is needed, not useless.

In other solutions like PHP, if you have 2 pages that both render some data, when you open the second page from the first page, that data will be refetched and users will see the latest value. 2 requests will be made for those 2 pages.

Same for SWR, 2 requests will be made and users will see the latest value when they navigate to the new page. The difference is that with the cache and stale-while-revalidate strategy, the page transition will be faster.

So as you can see, SWR isn't aimed to solve the "cache" problem of SPAs; instead it's trying to solve a bunch of UX problems.

However, if you really understand the downside of disabling this feature and you just want to use SWR as a cache tool (BTW in that case I'd suggest using a fetch lib with cache supported instead), you can turn off focus and reconnect revalidation and set a pretty large debounce interval.

**Comment by shuding** [comment]
Update: the alternative `revalidateOnCacheExists: false` option was landed in #992 (`useSWRImmutable` helper hook, and a `revalidateWhenStale` option), you can install `swr@beta` to try it out. We will update the docs when it's released in stable. 

Thanks for your input! ❤️

**Comment by Leyksnal** [comment]
Helpful
